src/class-controller.php

```diff
@@ -0,0 +1,215 @@
+<?php
+/**
+ * Controller class file
+ *
+ * @package wp_theme_migrator
+ */
+
+namespace Alley\WP\Theme_Migrator;
+
+use Alley\WP\Theme_Migrator\Context;
+use Alley\WP\Theme_Migrator\WP as WP_Theme_Migrator_WP;
+use Exception;
+
+/**
+ * Controller class.
+ */
+class Controller {
+	use \Alley\WP\Theme_Migrator\Theme_Helpers;
+
+	/**
+	 * Context instance.
+	 *
+	 * @var Context
+	 */
+	protected $context;
+
+	/**
+	 * WP_Theme_Migrator_WP instance.
+	 *
+	 * @var WP_Theme_Migrator_WP
+	 */
+	protected $wp;
+
+	/**
+	 * Whether to handle the current request with the theme we're migrating to.
+	 *
+	 * @var bool
+	 */
+	protected $should_migrate;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Context $context Context object.
+	 *
+	 * @throws Exception Thrown if context is not valid.
+	 */
+	public function __construct( Context $context ) {
+		if ( ! $context->is_valid_context() ) {
+			throw new Exception(
+				// @todo Add more robust error messaging.
+				sprintf(
+					__( 'WP Theme Migrator failed. Context is not valid.', 'wp-theme-migrator' ),
+				)
+			);
+		}
+
+		$this->context = $context;
+		$this->wp      = new WP_Theme_Migrator_WP();
+		$this->init();
+	}
+
+	/**
+	 * Add hooks to kick off migrator and tie in useful query vars.
+	 */
+	protected function init() {
+		// Post types and taxonomies must be registered by this point to be used
+		// for determining migratability.
+		add_action( 'setup_theme', [ $this, 'run' ], 100 );
+
+		// Add query vars for post types to our WP.
+		add_action( 'registered_post_type', [ $this, 'add_post_type_query_vars' ], 10, 2 );
```

Comment:
I like using the `action_{name_of_action}` convention for callbacks since it more clearly ties the hook to its callback. It's used elsewhere in this codebase and I'd recommend it be used here as well.

Reply:
@kevinfodness  I hear that. I like the convention for filters because ,with a filter, you're always going to be doing the same thing – returning modified data. When reading the `add_filter` line, it's clear what we're doing if we name the callback for the filter.

With an action though, anything can happen on that hook. I find it improves legibility at the `add_action` line to use the callback name to indicate what might happen. For example, it's clear in this line that when we register a post type, we're going to add the query vars. Then, when we add another action for `registered_taxonomy` to do something similar, the flow of actions is clear just in these few lines.